dropshot/src/pagination.rs

```diff
@@ -0,0 +1,569 @@
+ * behavior (e.g., a client lists projects in ascending order, but then asks for
+ * the next page in descending order).  The token also includes a version number
+ * so that it can be evolved in the future.
+ *
+ *
+ * ## Background: Why paginate HTTP APIs in the first place?
+ *
+ * Pagination helps ensure that the cost of a request in terms of resource
+ * utilization remains O(1) -- that is, it can be bounded above by a constant
+ * rather than scaling proportionally with any of the request parameters.  This
+ * simplifies utilization monitoring, capacity planning, and scale-out
+ * activities for the service, since operators can think of the service in terms
+ * of one unit that needs to be scaled up.  (It's still a very complex process,
+ * but it would be significantly harder if requests weren't O(1).)
+ *
+ * Similarly, pagination helps ensure that the time required for a request is
+ * O(1) under normal conditions.  This makes it easier to define expectations
+ * for service latency and to monitor that latency to determine if those
+ * expectations are violated.  Generally, if latency increases, then the service
+ * is unhealthy, and a crisp definition of "unhealthy" is important to operate a
+ * service with high availability.  If requests weren't O(1), an increase in
+ * latency might just reflect a changing workload that's still performing within
+ * expectations -- e.g., clients listing larger collections than they were
+ * before, but still getting results promptly.  That would make it much harder
+ * to see when the service really is unhealthy.
+ *
+ * Finally, bounding requests to O(1) work is a critical mitigation for common
+ * (if basic) denial-of-service (DoS) attacks because it requires that clients
+ * consume resources commensurate with the server costs that they're imposing.
+ * If a service exposes an API that does an unbounded amount of work
+ * proportional to some parameter, then it's cheap to launch a DoS on the
+ * service by just invoking that API with a large parameter.  By contrast, if
+ * the client has to do work that scales linearly with the work the server has
+ * to do, then the client's costs go up in order to scale up the attack.
+ *
+ * Along these lines, connections and requests consume finite server resources
+ * like open file descriptors and database connections.  If a service is built
+ * so that requests are all supposed to take about the same amount of time (or
+ * at least that there's a constant upper bound), then it may be possible to use
+ * a simple timeout scheme to cancel requests that are taking too long, as might
+ * happen if a malicious client finds some way to cause requests to hang or take
+ * a very long time.
+ *
+ * [1]: https://cloud.google.com/apis/design/design_patterns#list_pagination
+ * [2]: https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/
+ */
+
+use crate as dropshot; // XXX needed for ExtractedParameter below
```

Comment:
@ahl This was another thing I ran into that you might have an idea about.  I needed to use the `ExtractedParameter` proc macro in this file.  However, that macro spits out code that explicitly uses `dropshot` by name -- here's some `cargo expand` output:

```
    impl dropshot::ExtractedParameter for MarkerScanMode {
        fn metadata(
            _in: dropshot::ApiEndpointParameterLocation,
        ) -> Vec<dropshot::ApiEndpointParameter> {
            ::alloc::vec::Vec::new()
        }
    }
```

so that's why I've got `use crate as dropshot` here, but that feels weird.  Is there some other pattern for referring to this that doesn't require that the consumer has already `use`d the right thing?

Reply:
So I built something to handle this:
https://github.com/oxidecomputer/dropshot/blob/master/dropshot_endpoint/src/lib.rs#L142

Here's an examples of its use:
https://github.com/oxidecomputer/dropshot/blob/master/dropshot/src/api_description.rs#L878

I think you can do e.g.
```rust
#[derive(Debug, Deserialize, ExtractedParameter)]
#[serde(rename_all = "kebab-case")]
#[dropshot(crate = crate)]
pub enum MarkerScanMode {
    ByMarkerAsc,
}
```

It's going to be a little annoying to do for all uses. As I've mentioned, I think `ExtractedParameter` is not going to end up being the right solution (e.g. as we handle some of the stuff it's punting on now, non-required parameters)

Reply:
Ah okay.  Right now it feels like `use crate as dropshot` is a cleaner workaround for this file -- does that seem reasonable?

I'd hoped [`$crate`](https://doc.rust-lang.org/1.5.0/book/macros.html#the-variable-crate) could help here, but when I tried it, I ran into the behavior [described here](https://stackoverflow.com/questions/48050155/is-it-possible-to-get-the-full-namespace-of-a-struct-in-a-custom-derive#comment83076899_48052098).

Reply:
Yes; it feels a little janky, but less annoying than sprinkling that magic everywhere..